larsim/GenericCRT/GenericCRT.cxx

```diff
@@ -0,0 +1,150 @@
+/*!
+ * Title:   GenericCRT Utility Class
+ * Author:  Andrzej Szelc (andrzejs@fnal.gov)
+ *
+ * Description:
+ * Class with Algorithms to convert sim::AuxDetHits to sim::AuxDetSimChannels
+ *
+*/
+
+
+#include "GenericCRT.h"
+
+#include <utility> // std::move()
+#include <algorithm> // std::find()
+
+
+
+
+// sim::GenericCRTUtility::GenericCRTUtility()
+// {
+//
+//
+// }
+
+
+
+unsigned int sim::GenericCRTUtility::CopyAuxDetHittoAuxDetIDE(const sim::AuxDetHit &InputHit,sim::AuxDetIDE &outputIDE)
+{
+
+
+   outputIDE.trackID		    = InputHit.GetTrackID();
+   outputIDE.energyDeposited	= InputHit.GetEnergyDeposited();
+   outputIDE.entryX		= InputHit.GetEntryX();
+   outputIDE.entryY		= InputHit.GetEntryY();
+   outputIDE.entryZ		= InputHit.GetEntryZ();
+   outputIDE.entryT		= InputHit.GetEntryT();
+   outputIDE.exitX		= InputHit.GetExitX();
+   outputIDE.exitY		= InputHit.GetExitY();
+   outputIDE.exitZ		= InputHit.GetExitZ();
+   outputIDE.exitT		= InputHit.GetExitT();
+   outputIDE.exitMomentumX	= InputHit.GetExitMomentumX();
+   outputIDE.exitMomentumY	= InputHit.GetExitMomentumY();
+   outputIDE.exitMomentumZ	= InputHit.GetExitMomentumZ();
+
+
+   return  InputHit.GetID();
+}
+
+
+
+
+unsigned int sim::GenericCRTUtility::GetNumberofAuxDetChannels(const std::vector<sim::AuxDetHit> &InputHitVector,std::vector<unsigned int> &AuxDetChanNumbers )
+{
+
+  AuxDetChanNumbers.reserve(InputHitVector.size());
+
+
+  for(auto HitIter : InputHitVector)
+  {
+
+  std::vector<unsigned int>::iterator Chanitr
+      = std::find(AuxDetChanNumbers.begin(), AuxDetChanNumbers.end(), HitIter.GetID());
+
+   if(Chanitr == AuxDetChanNumbers.end()){ //If trackID is already in the map, update it
+         //if channel ID is not in the set yet, add it
+      AuxDetChanNumbers.push_back(HitIter.GetID());
+        }//
+
+  }
+
+return AuxDetChanNumbers.size();
+
+}
+
+
+
+sim::AuxDetSimChannel const sim::GenericCRTUtility::GetAuxDetSimChannelByNumber( const std::vector<sim::AuxDetHit> &InputHitVector,unsigned int inputchannel)
+{
+        std::vector<sim::AuxDetIDE> IDEvector;
+       IDEvector.resize(0);
+        //loop over sim::AuxDetHits and assign them to AuxDetSimChannels.
+
+        size_t ad_id_no = 9999;
+        size_t ad_sen_id_no = 9999;
+
+        for(auto AuxDetHitIter : InputHitVector)
+        {
+
+            unsigned int channel = AuxDetHitIter.GetID();
+            double xcoordinate = (AuxDetHitIter.GetEntryX() + AuxDetHitIter.GetExitX())/2.0;
+            double ycoordinate = (AuxDetHitIter.GetEntryY() + AuxDetHitIter.GetExitY())/2.0;
+            double zcoordinate = (AuxDetHitIter.GetEntryZ() + AuxDetHitIter.GetExitZ())/2.0;
+            double worldPos[3] = {xcoordinate,ycoordinate,zcoordinate};
+            fGeo->FindAuxDetSensitiveAtPosition(worldPos, ad_id_no, ad_sen_id_no, 0.0001);
+            if(channel == inputchannel)   // this is the channel we want.
+            {
```

Comment:
A few issues:

- `AuxDetHitIter` is not an iterator, but it is an `sim::AuxDetHit` object.  To avoid, copying each hit object during iteration, use a reference.
- You only want to calculate the x,y,z-coordinates for the channel you care about.  Otherwise you're wasting a lot of computing time.

I'd therefore restructure the loop like:

```c++
for (auto const& auxDetHit : InputHitVector) {
  // Skip all channels that do not match the input channel
  if (auxDetHit.GetID() != inputchannel) continue;

  double xcoordinate = ...;
  // ...
}
```

Reply:
Changes made